x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/accesscontrol/OptOutQueryCache.java

```diff
@@ -25,69 +25,42 @@
- * invoked on this query cache.
+ * Opts out of the query cache if field level security is active for the current request, and it is unsafe to cache.
  */
-public final class OptOutQueryCache extends AbstractIndexComponent implements LicenseStateListener, QueryCache {
+public final class OptOutQueryCache extends AbstractIndexComponent implements QueryCache {
 
     private final IndicesQueryCache indicesQueryCache;
     private final ThreadContext context;
     private final String indexName;
-    private final XPackLicenseState licenseState;
-    private volatile boolean licenseStateListenerRegistered;
+    private final Settings settings;
 
     public OptOutQueryCache(
             final IndexSettings indexSettings,
             final IndicesQueryCache indicesQueryCache,
             final ThreadContext context,
-            final XPackLicenseState licenseState) {
+            final Settings settings) {
         super(indexSettings);
         this.indicesQueryCache = indicesQueryCache;
         this.context = Objects.requireNonNull(context, "threadContext must not be null");
         this.indexName = indexSettings.getIndex().getName();
-        this.licenseState = Objects.requireNonNull(licenseState, "licenseState");
-    }
-
-    /**
-     * Register this query cache to listen for license state changes. This must be done after construction of this query cache before any
-     * other public methods are invoked on this query cache.
-     */
-    public void listenForLicenseStateChanges() {
-        /*
-         * Registering this as a listener can not be done in the constructor because otherwise it would be unsafe publication of this. That
-         * is, it would expose this to another thread before the constructor had finished. Therefore, we have a dedicated method to register
-         * the listener that is invoked after the constructor has returned.
-         */
-        assert licenseStateListenerRegistered == false;
-        licenseState.addListener(this);
-        licenseStateListenerRegistered = true;
+        this.settings = Objects.requireNonNull(settings, "settings must not be null");
     }
 
     @Override
     public void close() throws ElasticsearchException {
-        assert licenseStateListenerRegistered;
-        licenseState.removeListener(this);
         clear("close");
     }
 
-    @Override
-    public void licenseStateChanged() {
-        assert licenseStateListenerRegistered;
-        clear("license state changed");
-    }
 
     @Override
     public void clear(final String reason) {
-        assert licenseStateListenerRegistered;
         logger.debug("full cache clear, reason [{}]", reason);
         indicesQueryCache.clearIndex(index().getName());
     }
 
     @Override
     public Weight doCache(Weight weight, QueryCachingPolicy policy) {
-        assert licenseStateListenerRegistered;
-        if (licenseState.isSecurityEnabled() == false) {
+        if (XPackSettings.SECURITY_ENABLED.get(settings) == false) {
```

Comment:
I don't think this is necessary. `Security.onIndexModule` already skips installing this component if security is disabled.